src/lib/storage/dictionary_segment.cpp

```diff
@@ -16,15 +16,78 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
                                         const std::shared_ptr<const BaseCompressedVector>& attribute_vector)
     : BaseDictionarySegment(data_type_from_type<T>()),
       _dictionary{dictionary},
+      _dictionary_span{std::make_shared<std::span<const T>>(_dictionary->data(), _dictionary->size())},
       _attribute_vector{attribute_vector},
       _decompressor{_attribute_vector->create_base_decompressor()} {
   // NULL is represented by _dictionary.size(). INVALID_VALUE_ID, which is the highest possible number in
   // ValueID::base_type (2^32 - 1), is needed to represent "value not found" in calls to lower_bound/upper_bound.
   // For a DictionarySegment of the max size Chunk::MAX_SIZE, those two values overlap.
-
   Assert(_dictionary->size() < std::numeric_limits<ValueID::base_type>::max(), "Input segment too big");
 }
 
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const std::span<const T>>& dictionary,
+                                        const std::shared_ptr<const BaseCompressedVector>& attribute_vector)
+    : BaseDictionarySegment(data_type_from_type<T>()),
+      _dictionary{},
+      _dictionary_span{dictionary},
+      _attribute_vector{attribute_vector},
+      _decompressor{_attribute_vector->create_base_decompressor()} {
+  // NULL is represented by _dictionary.size(). INVALID_VALUE_ID, which is the highest possible number in
+  // ValueID::base_type (2^32 - 1), is needed to represent "value not found" in calls to lower_bound/upper_bound.
+  // For a DictionarySegment of the max size Chunk::MAX_SIZE, those two values overlap.
+  Assert(_dictionary_span->size() < std::numeric_limits<ValueID::base_type>::max(), "Input segment too big");
+}
+
+template <typename T>
+std::shared_ptr<DictionarySegment<T>> DictionarySegment<T>::create(const uint32_t* map,
+                                                                   const uint32_t segment_start_offset_bytes) {
+  if constexpr (std::is_same_v<T, int32_t>) {
+    const auto segment_start_map_index = segment_start_offset_bytes / 4;
+    const auto dictionary_size = map[segment_start_map_index];
+    const auto attribute_vector_size = map[segment_start_map_index + 1];
+    const auto encoding_type = map[segment_start_map_index + 2];
+    const auto type_size_as_index = sizeof(T) / 4;
+
+    auto* dictionary_map = reinterpret_cast<const T*>(map);
+    auto dictionary_span = std::span<const T>(&dictionary_map[segment_start_map_index + 3], dictionary_size);
+    auto dictionary_span_pointer = std::make_shared<std::span<const T>>(dictionary_span);
+
+    if (encoding_type == 1) {
```

Comment:
Better instantiate a real `EncodingType` and use a switch.

Reply:
Something similar already exists in `StorageManager.hpp`, might be interesting